packages/flutter/lib/src/material/text_field.dart

```diff
@@ -461,6 +468,16 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
         hintMaxLines: widget.decoration?.hintMaxLines ?? widget.maxLines
       );
 
+    // No need to build anything if counter or counterText were given directly.
+    if (effectiveDecoration.counter != null
+        || effectiveDecoration.counterText != null)
+      return effectiveDecoration;
+
+    if (effectiveDecoration.counterText != null) {
+      return effectiveDecoration;
+    }
+
+    // If buildCounter was provided, use it to generate a counter widget.
     Widget counter;
     final int currentLength = _effectiveController.value.text.runes.length;
     if (effectiveDecoration.counter == null
```

Comment:
Whoops, fixed the redundant conditional and put the first one on one line.

The logic should be correct as it is.  If `counter` is specified then it can be used as the counter widget directly.  If `counterText` is specified then input decorator will build a widget for the given text.  If neither are specified then this method might need to build a counter widget.